contracts/epochs/FlowDKG.cdc

```diff
@@ -0,0 +1,251 @@
+                    return
+                }
+
+                // Check each key in the submiission to make sure that it matches
+                // the existing one
+                for key in submission {
+                    if key.length != 128 {
+                        // correct key length?
+                        return
+                    }
+                    // if a key is different, stop checking this submission
+                    // and move on to the next one
+                    if key != existingSubmission[index] {
+                        break
+                    }
+
+                    index = index + 1
+                }
+
+                // If we have gotten to the last key and they have all matched
+                // update the counter for this submission and emit the event
+                if index == submission.length {
+                    FlowDKG.uniqueFinalSubmissionCount[finalSubmissionIndex] = FlowDKG.uniqueFinalSubmissionCount[finalSubmissionIndex]! + UInt64(1)
+                    emit FinalSubmission(nodeID: self.nodeID, submission: submission)
+                    break
+                }
+
+                finalSubmissionIndex = finalSubmissionIndex + 1
+
+                // If no matches were found, add this submission as a new unique one
+                // and emit an event
+                if finalSubmissionIndex == FlowDKG.uniqueFinalSubmissions.length {
+                    FlowDKG.uniqueFinalSubmissionCount[finalSubmissionIndex] = 0
+                    FlowDKG.uniqueFinalSubmissions.append(submission)
+                    emit FinalSubmission(nodeID: self.nodeID, submission: submission)
+                }
+            }
+        }
+
+        init(nodeID: String) {
+            pre {
+                FlowDKG.participantIsRegistered(nodeID): "Cannot create a Participant for a node ID that hasn't been registered"
+                !FlowDKG.participantIsClaimed(nodeID)!: "Cannot create a Participant resource for a node ID that has already been claimed"
+            }
+            self.nodeID = nodeID
+            FlowDKG.nodeClaimed[nodeID] = true
+        }
+
+        destroy () {
+            FlowDKG.nodeClaimed[self.nodeID] = false
+        }
+
+    }
+
+    // The Admin resource provides the ability to begin and end voting for an epoch
+    pub resource Admin {
+
+        /// Creates a new Participant resource for a consensus node
+        pub fun createParticipant(nodeID: String): @Participant {
+            let participant <-create Participant(nodeID: nodeID)
+            FlowDKG.nodeClaimed[nodeID] = true
+            return <-participant
+        }
+
+        /// Resets all the fields for tracking the current DKG process
+        /// and sets the given node IDs as registered
+        pub fun startDKG(nodeIDs: [String]) {
+            FlowDKG.currentPhase = DKGPhase.Phase1
+
+            for id in nodeIDs {
+                FlowDKG.nodeHasSubmitted[id] = false
+            }
```

Comment:
I think we also need to reset existing values in this mapping. Otherwise DKG participants from last epoch that aren't DKG participants in this epoch would still be considered valid.
```suggestion
            FlowDKG.nodeHasSubmitted = {}
            for id in nodeIDs {
                FlowDKG.nodeHasSubmitted[id] = false
            }
```